File: backend/add_sample_data.js
const sqlite3 = require('sqlite3').verbose();
const moment = require('moment');

// Database setup
const db = new sqlite3.Database('./smart_tank.db');

console.log('Adding sample data to Smart Tank database...');

// Add device
db.run(`INSERT OR IGNORE INTO devices (device_id, tank_height_cm, tank_diameter_cm, tank_capacity_liters, leak_threshold_percent, overflow_threshold_percent)
        VALUES (?, ?, ?, ?, ?, ?)`,
       ['tank01', 150.0, 100.0, 1178.0, 1.0, 95.0], function(err) {
  if (err) {
    console.error('Error adding device:', err);
  } else {
    console.log('Added device tank01');
  }
});

// Add sample telemetry data for the last 24 hours
const now = moment();
const baseLevel = 112.5; // 75% of 150cm

for (let i = 0; i < 24 * 4; i++) { // Every 15 minutes for 24 hours
  const timestamp = now.clone().subtract(15 * i, 'minutes').toDate();

  // Simulate gradual water consumption
  const levelDrop = i * 0.5;
  const currentLevel = Math.max(20, baseLevel - levelDrop + (Math.random() - 0.5) * 4);

  const percentFull = (currentLevel / 150.0) * 100;

  // Simulate pump turning on when level gets low
  const pumpState = percentFull < 30 ? 'ON' : 'OFF';

  // Simulate flow rate
  const flowRate = pumpState === 'OFF' ? Math.random() * 5 : 10 + Math.random() * 10;

  db.run(`INSERT INTO telemetry (device_id, timestamp, level_cm, tank_height_cm, percent_full, flow_l_min, pump_state, temperature_c, tds_ppm)
          VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`,
         ['tank01', timestamp, currentLevel, 150.0, percentFull, flowRate, pumpState,
          20 + Math.random() * 5, 200 + Math.random() * 200]);
}

console.log(`Added ${24 * 4} telemetry records`);

// Close database
db.close((err) => {
  if (err) {
    console.error('Error closing database:', err);
  } else {
    console.log('Sample data added successfully!');
    console.log('Please restart the backend server to load the new data.');
  }
});

File: backend/add_sample_data.py
#!/usr/bin/env python3
"""
Add sample data to the Smart Tank database for testing
"""

import sys
import os
from datetime import datetime, timedelta
import random

# Add the backend directory to the path
sys.path.append(os.path.dirname(os.path.abspath(__file__)))

from main import SessionLocal, Device, Telemetry, device_status
from sqlalchemy.orm import Session

def add_sample_device(db: Session):
    """Add a sample device to the database"""
    # Check if device already exists
    existing_device = db.query(Device).filter(Device.device_id == "tank01").first()
    if existing_device:
        print("Device tank01 already exists")
        return existing_device
    
    device = Device(
        device_id="tank01",
        tank_height_cm=150.0,
        tank_diameter_cm=100.0,
        tank_capacity_liters=1178.0,
        leak_threshold_percent=1.0,
        overflow_threshold_percent=95.0,
        created_at=datetime.utcnow()
    )
    
    db.add(device)
    db.commit()
    print("Added device tank01")
    return device

def add_sample_telemetry(db: Session):
    """Add sample telemetry data"""
    # Generate sample data for the last 24 hours
    now = datetime.utcnow()
    
    # Start with a tank that's 75% full and gradually decreasing
    base_level = 112.5  # 75% of 150cm
    
    for i in range(24 * 4):  # Every 15 minutes for 24 hours
        timestamp = now - timedelta(minutes=15 * i)
        
        # Simulate gradual water consumption
        level_drop = i * 0.5  # Gradual decrease
        current_level = max(20, base_level - level_drop + random.uniform(-2, 2))
        
        percent_full = (current_level / 150.0) * 100
        
        # Simulate pump turning on when level gets low
        pump_state = "ON" if percent_full < 30 else "OFF"
        
        # Simulate flow rate
        flow_rate = random.uniform(0, 5) if pump_state == "OFF" else random.uniform(10, 20)
        
        telemetry = Telemetry(
            device_id="tank01",
            timestamp=timestamp,
            level_cm=current_level,
            tank_height_cm=150.0,
            percent_full=percent_full,
            flow_l_min=flow_rate,
            pump_state=pump_state,
            temperature_c=random.uniform(20, 25),
            tds_ppm=random.uniform(200, 400)
        )
        
        db.add(telemetry)
    
    db.commit()
    print(f"Added {24 * 4} telemetry records")

def update_device_status():
    """Update the in-memory device status"""
    db = SessionLocal()
    try:
        # Get the latest telemetry
        latest = db.query(Telemetry).filter(
            Telemetry.device_id == "tank01"
        ).order_by(Telemetry.timestamp.desc()).first()
        
        if latest:
            device_status["tank01"] = {
                "current_level": latest.level_cm,
                "percent_full": latest.percent_full,
                "flow_rate": latest.flow_l_min,
                "pump_state": latest.pump_state == "ON",
                "temperature": latest.temperature_c,
                "tds": latest.tds_ppm,
                "last_update": latest.timestamp,
                "days_until_empty": 5.2,  # Sample prediction
                "consumption_today": 45.5,  # Sample consumption
                "consumption_week": 320.0   # Sample weekly consumption
            }
            print("Updated device status in memory")
    finally:
        db.close()

def main():
    """Main function to add sample data"""
    print("Adding sample data to Smart Tank database...")
    
    db = SessionLocal()
    try:
        # Add device
        add_sample_device(db)
        
        # Add telemetry data
        add_sample_telemetry(db)
        
        print("Sample data added successfully!")
        
    except Exception as e:
        print(f"Error adding sample data: {e}")
        db.rollback()
    finally:
        db.close()
    
    # Update in-memory status
    update_device_status()

if __name__ == "__main__":
    main()